pkg/asset/upgrade/bundle.go

```diff
@@ -0,0 +1,489 @@
+package upgrade
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/bombsimon/logrusr/v4"
+	"github.com/distribution/distribution/v3/reference"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/openshift/appliance/pkg/asset/config"
+	"github.com/openshift/appliance/pkg/jq"
+	"github.com/openshift/appliance/pkg/registry"
+	"github.com/openshift/installer/pkg/asset"
+)
+
+// Bundle generates the upgrade bundle.
+type Bundle struct {
+	Tar    *asset.File
+	Digest *asset.File
+}
+
+var _ asset.Asset = (*Bundle)(nil)
+
+// Name returns a human friendly name for the asset.
+func (b *Bundle) Name() string {
+	return "Upgrade bundle"
+}
+
+// Dependencies returns all the dependencies directly needed to generate the asset.
+func (b *Bundle) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&config.EnvConfig{},
+		&config.ApplianceConfig{},
+	}
+}
+
+// Generate generates the upgrade bundle.
+func (b *Bundle) Generate(dependencies asset.Parents) error {
+	// Get the logger:
+	logger := logrus.StandardLogger()
+
+	// Get the dependencies:
+	envCfg := &config.EnvConfig{}
+	appCfg := &config.ApplianceConfig{}
+	dependencies.Get(envCfg, appCfg)
+
```

Comment:
Is there anything worth saving in cache dir? If so, need to first check if already exists using envConfig.FindInCache.